wasm error running example

Running the host-functions example I get the following error:

```
2022/11/12 14:42:05 Sending a HTTP request...
2022/11/12 14:42:05 wasm error: unreachable
wasm stack trace:
	.runtime.runtimePanic(i32,i32)
	.runtime.lookupPanic()
	.malloc(i32) i32 (recovered by wazero)
wasm stack trace:
	env.log(i32,i32) i64
	.greeter_greet(i32,i32) i64
```

* go version go1.18.7 darwin/arm64
* macOS 12.6.1 
* Chip M1 Max

Hey, I have the same problem with loading my plugin

- MacOS: `13.0.1`
- Chip: `Apple M1 Max`
- Go: `go1.19.2 darwin/arm64`
- go-plugin: `0.2.0`

```
[2022/11/17T19:29:43] [application] [ERROR] [core.go:82,func1] wasm error: unreachable
wasm stack trace:
	.runtime.runtimePanic(i32,i32)
	.runtime.lookupPanic()
	.malloc(i32) i32  
```

--- a/.github/workflows/test.yaml
+++ b/.github/workflows/test.yaml
@@ -11,11 +11,19 @@ on:
 env:
   GO_VERSION: "1.18"
   PROTOC_VERSION: "21.5"
-  TINYGO_VERSION: "0.24.0"
+
 jobs:
   test:
     name: Test
     runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        tinygo-version:
+          - "0.24.0"
+          - "0.25.0"
+          - "0.26.0"
+
     steps:
       - uses: actions/checkout@v3
 
@@ -41,8 +49,8 @@ jobs:
 
       - name: Install TinyGo
         run: |
-          wget https://github.com/tinygo-org/tinygo/releases/download/v${TINYGO_VERSION}/tinygo_${TINYGO_VERSION}_amd64.deb
-          sudo dpkg -i tinygo_${TINYGO_VERSION}_amd64.deb
+          wget https://github.com/tinygo-org/tinygo/releases/download/v${{ matrix-tinygo-version }}/tinygo_${{ matrix-tinygo-version }}_amd64.deb
+          sudo dpkg -i tinygo_${{ matrix-tinygo-version }}_amd64.deb
 
       - name: Run unit tests
         run: make test

--- a/.github/workflows/test.yaml
+++ b/.github/workflows/test.yaml
@@ -14,7 +14,7 @@ env:
 
 jobs:
   test:
-    name: Test
+    name: Test (TinyGo ${{ matrix.tinygo-version }})
     runs-on: ubuntu-latest
     strategy:
       fail-fast: false

--- a/.github/workflows/test.yaml
+++ b/.github/workflows/test.yaml
@@ -49,8 +49,8 @@ jobs:
 
       - name: Install TinyGo
         run: |
-          wget https://github.com/tinygo-org/tinygo/releases/download/v${{ matrix-tinygo-version }}/tinygo_${{ matrix-tinygo-version }}_amd64.deb
-          sudo dpkg -i tinygo_${{ matrix-tinygo-version }}_amd64.deb
+          wget https://github.com/tinygo-org/tinygo/releases/download/v${{ matrix.tinygo-version }}/tinygo_${{ matrix.tinygo-version }}_amd64.deb
+          sudo dpkg -i tinygo_${{ matrix.tinygo-version }}_amd64.deb
 
       - name: Run unit tests
         run: make test

--- a/examples/helloworld/greeting/greet_host.pb.go
+++ b/examples/helloworld/greeting/greet_host.pb.go
@@ -129,6 +129,9 @@ func (p *greeterPlugin) Greet(ctx context.Context, request GreetRequest) (respon
 		return response, err
 	}
 	dataSize := uint64(len(data))
+	if dataSize == 0 {
+		return response, nil
+	}
 	results, err := p.malloc.Call(ctx, dataSize)
 	if err != nil {
 		return response, err

--- a/examples/host-functions/greeting/greet_host.pb.go
+++ b/examples/host-functions/greeting/greet_host.pb.go
@@ -221,6 +221,9 @@ func (p *greeterPlugin) Greet(ctx context.Context, request GreetRequest) (respon
 		return response, err
 	}
 	dataSize := uint64(len(data))
+	if dataSize == 0 {
+		return response, nil
+	}
 	results, err := p.malloc.Call(ctx, dataSize)
 	if err != nil {
 		return response, err

--- a/examples/known-types/known/known_host.pb.go
+++ b/examples/known-types/known/known_host.pb.go
@@ -129,6 +129,9 @@ func (p *wellKnownPlugin) Diff(ctx context.Context, request DiffRequest) (respon
 		return response, err
 	}
 	dataSize := uint64(len(data))
+	if dataSize == 0 {
+		return response, nil
+	}
 	results, err := p.malloc.Call(ctx, dataSize)
 	if err != nil {
 		return response, err

--- a/examples/wasi/cat/cat_host.pb.go
+++ b/examples/wasi/cat/cat_host.pb.go
@@ -129,6 +129,9 @@ func (p *fileCatPlugin) Cat(ctx context.Context, request FileCatRequest) (respon
 		return response, err
 	}
 	dataSize := uint64(len(data))
+	if dataSize == 0 {
+		return response, nil
+	}
 	results, err := p.malloc.Call(ctx, dataSize)
 	if err != nil {
 		return response, err

--- a/gen/host.go
+++ b/gen/host.go
@@ -289,7 +289,9 @@ func genPluginMethod(g *protogen.GeneratedFile, f *fileInfo, method *protogen.Me
 	g.P("data, err := request.MarshalVT()")
 	g.P(errorHandling)
 	g.P("dataSize := uint64(len(data))")
-
+	g.P(`if dataSize == 0 {
+			return response, nil 
+		}`)
 	g.P("results, err := p.malloc.Call(ctx, dataSize)")
 	g.P(errorHandling)
 

--- a/tests/fields/proto/fields_host.pb.go
+++ b/tests/fields/proto/fields_host.pb.go
@@ -129,6 +129,9 @@ func (p *fieldTestPlugin) Test(ctx context.Context, request Request) (response R
 		return response, err
 	}
 	dataSize := uint64(len(data))
+	if dataSize == 0 {
+		return response, nil
+	}
 	results, err := p.malloc.Call(ctx, dataSize)
 	if err != nil {
 		return response, err

--- a/tests/host-functions/proto/host_host.pb.go
+++ b/tests/host-functions/proto/host_host.pb.go
@@ -185,6 +185,9 @@ func (p *greeterPlugin) Greet(ctx context.Context, request GreetRequest) (respon
 		return response, err
 	}
 	dataSize := uint64(len(data))
+	if dataSize == 0 {
+		return response, nil
+	}
 	results, err := p.malloc.Call(ctx, dataSize)
 	if err != nil {
 		return response, err

--- a/tests/import/proto/bar/bar_host.pb.go
+++ b/tests/import/proto/bar/bar_host.pb.go
@@ -129,6 +129,9 @@ func (p *barPlugin) Hello(ctx context.Context, request Request) (response Reply,
 		return response, err
 	}
 	dataSize := uint64(len(data))
+	if dataSize == 0 {
+		return response, nil
+	}
 	results, err := p.malloc.Call(ctx, dataSize)
 	if err != nil {
 		return response, err

--- a/tests/import/proto/foo/foo_host.pb.go
+++ b/tests/import/proto/foo/foo_host.pb.go
@@ -130,6 +130,9 @@ func (p *fooPlugin) Hello(ctx context.Context, request Request) (response bar.Re
 		return response, err
 	}
 	dataSize := uint64(len(data))
+	if dataSize == 0 {
+		return response, nil
+	}
 	results, err := p.malloc.Call(ctx, dataSize)
 	if err != nil {
 		return response, err

--- a/tests/well-known/proto/known_host.pb.go
+++ b/tests/well-known/proto/known_host.pb.go
@@ -130,6 +130,9 @@ func (p *knownTypesTestPlugin) Test(ctx context.Context, request Request) (respo
 		return response, err
 	}
 	dataSize := uint64(len(data))
+	if dataSize == 0 {
+		return response, nil
+	}
 	results, err := p.malloc.Call(ctx, dataSize)
 	if err != nil {
 		return response, err
@@ -275,6 +278,9 @@ func (p *emptyTestPlugin) DoNothing(ctx context.Context, request emptypb.Empty) 
 		return response, err
 	}
 	dataSize := uint64(len(data))
+	if dataSize == 0 {
+		return response, nil
+	}
 	results, err := p.malloc.Call(ctx, dataSize)
 	if err != nil {
 		return response, err

--- a/wasm/host.go
+++ b/wasm/host.go
@@ -25,7 +25,13 @@ func WriteMemory(ctx context.Context, m api.Module, data []byte) (uint64, error)
 	if malloc == nil {
 		return 0, errors.New("malloc is not exported")
 	}
-	results, err := malloc.Call(ctx, uint64(len(data)))
+
+	l := uint64(len(data))
+	if l == 0 {
+		return 0, nil
+	}
+
+	results, err := malloc.Call(ctx, l)
 	if err != nil {
 		return 0, err
 	}

